### PR TITLE
DataViews:  Fix `array` field type validation for empty values

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Extracted from https://github.com/WordPress/gutenberg/pull/81377

The built-in validation of the `array` field type crashed on empty values (`undefined`, `''`, `null`) and the exception surfaced as a validation error, making forms with an empty optional array field invalid. This was also inconsistent with the other field types, whose built-in validation lets empty values pass. This PR allows empty values — non-empty values should be enforced with the `required` rule instead.

Note: as pointed out in [review](https://github.com/WordPress/gutenberg/pull/81378#pullrequestreview-4898066609), this doesn't cover fields that declare `elements` — the `elements` rule is enabled by default and `isValidElements` does `[].concat( value )`, so an empty value still fails as `[ undefined ]`. That behavior predates this PR and is shared by all field types, so I'd rather handle it in a follow-up. This fix applies to array fields without `elements` (or with `isValid.elements: false`).

## Testing Instructions

There are unit tests for both the empty (valid) and the non-string values (invalid) cases. You can also test in the validation story, where the `Array` field is empty on load, by disabling the other fields' validation: http://localhost:50240/?path=/story/dataviews-dataform--validation&args=required:!false;elements:none;custom:none

1. `npm run storybook:dev` and open the above link.
2. In trunk the `Submit` button is disabled on load. With this PR it's enabled.

### Use of AI Tools
Generated with Fable 5 and adjusted/reviewed manually.

### CODE REVIEW

**Review by jorgefilipecosta — APPROVED** [review]
Hi @ntsekouras, thank you for the fix 👍

I think there is a case where an empty value is still reported as invalid: when the array field declares elements, the elements rule is enabled by default and runs before the custom validator, and isValidElements does [].concat( value ).every( ... ), so an empty value becomes [ undefined ] and fails with "Value must be one of the elements.". So for an optional array field with elements the form stays invalid when the value is empty (undefined, null and '' all fail), and in that case it seems the required rule is not what gates emptiness.

I used this story to reproduce it, the field is optional and empty but isValid is false and the submit button stays disabled:

```
diff --git a/packages/dataviews/src/dataform/stories/index.story.tsx b/packages/dataviews/src/dataform/stories/index.story.tsx
index 0c3ebc05e55..5df0ddcc08b 100644
--- a/packages/dataviews/src/dataform/stories/index.story.tsx
+++ b/packages/dataviews/src/dataform/stories/index.story.tsx
@@ -5,9 +5,11 @@ import LayoutMixedComponent from './layout-mixed';
 import LayoutRegularComponent from './layout-regular';
 import LayoutRowComponent from './layout-row';
 import LayoutPanelComponent from './layout-panel';
 import DataAdapterComponent from './data-adapter';
-import ValidationComponent from './validation';
+import ValidationComponent, {
+	ValidationEmptyArrayComponent,
+} from './validation';
 import VisibilityComponent from './visibility';
 
 const meta = {
 	tags: [ 'manifest' ],
@@ -170,8 +172,12 @@ export const Validation = {
 		minMax: false,
 	},
 };
 
+export const ValidationEmptyArray = {
+	render: ValidationEmptyArrayComponent,
+};
+
 export const Visibility = {
 	render: VisibilityComponent,
 };
 
diff --git a/packages/dataviews/src/dataform/stories/validation.tsx b/packages/dataviews/src/dataform/stories/validation.tsx
index 1ca43f69be5..6dc49799350 100644
--- a/packages/dataviews/src/dataform/stories/validation.tsx
+++ b/packages/dataviews/src/dataform/stories/validation.tsx
@@ -1079,5 +1079,73 @@ const ValidationComponent = ( {
 		</form>
 	);
 };
 
+type EmptyArrayItem = {
+	countries?: string[];
+};
+
+const emptyArrayFields: Field< EmptyArrayItem >[] = [
+	{
+		id: 'countries',
+		label: 'Countries (optional)',
+		type: 'array',
+		placeholder: 'Select countries',
+		description:
+			'Optional array field with elements. The initial empty value should be valid, but the elements rule reports it as invalid.',
+		elements: [
+			{ value: 'us', label: 'United States' },
+			{ value: 'ca', label: 'Canada' },
+			{ value: 'uk', label: 'United Kingdom' },
+		],
+	},
+];
+
+const emptyArrayForm = { fields: [ 'countries' ] };
+
+export const ValidationEmptyArrayComponent = () => {
+	const [ post, setPost ] = useState< EmptyArrayItem >( {
+		countries: undefined,
+	} );
+
+	const { validity, isValid } = useFormValidity(
+		post,
+		emptyArrayFields,
+		emptyArrayForm
+	);
+
+	return (
+		<form>
+			<Stack direction="column" align="start" gap="3xl">
+				<DataForm< EmptyArrayItem >
+					data={ post }
+					fields={ emptyArrayFields }
+					form={ emptyArrayForm }
+					validity={ validity }
+					onChange={ ( edits ) =>
+						setPost( ( prev ) => ( {
+							...prev,
+							...edits,
+						} ) )
+					}
+				/>
+				<p>
+					{ `isValid: ${ String( isValid ) } — countries validity: ${
+						validity?.countries
+							? JSON.stringify( validity.countries )
+							: 'valid'
+					}` }
+				</p>
+				<Button
+					__next40pxDefaultSize
+					accessibleWhenDisabled
+					disabled={ ! isValid }
+					variant="primary"
+				>
+					Submit
+				</Button>
+			</Stack>
+		</form>
+	);
+};
+
 export default ValidationComponent;
```


I guess isValidElements could also treat empty values as valid, but that change affects all field types with elements, so probably it is fine to handle it in a follow-up.


This fixes in the non elements case so I think it is good to merged even without the elements follow up.
